src/levanter/inference/engine.py

```diff
@@ -778,31 +866,52 @@ def _pages_needed_for_prompt(self, prompt_len: int) -> int:
     def _prefill_prompts(
         self,
         requests: Sequence[Request],
-    ) -> TokenQueue | None:
-        """Assign slot ids, set per-seq params, and build a single TokenQueue for prefill.
+    ) -> PrefillWork | None:
+        """Pack prompt work into a single PrefillWork structure for downstream device execution."""
 
-        Does not run device prefill; caller should invoke _run_prefill once with the returned queue.
-        """
-        max_seqs_in_prefill = int(self.config.max_seqs_in_prefill)
+        decode_state = self.gen_state.decode_state
+        max_seqs_in_prefill = self.config.max_seqs_in_prefill
         max_prefill_size = self.config.max_prefill_size or self.model.Pos.size
-        tokens = np.full((max_prefill_size,), INVALID, dtype=jnp.int32)
-        slot_ids = np.full((max_prefill_size,), INVALID, dtype=jnp.int32)
-        pos_ids = np.full((max_prefill_size,), INVALID, dtype=jnp.int32)
-        offset = 0
-        num_seqs_in_prefill = 0
+        max_seq_len = decode_state.tokens.axis_size("position")
+        max_slots = decode_state.max_seqs
+
+        queue_tokens = np.full((max_prefill_size,), INVALID, dtype=jnp.int32)
```

Comment:
nit the `np` api with `jnp.int32` is a little odd looking, maybe a convention i'm not aware of though

Reply:
oh yeah. will fix